cpp/src/cluster/detail/mst.cuh

```diff
@@ -117,6 +120,143 @@ void connect_knn_graph(
   merge_msts<value_idx, value_t>(msf, new_mst, stream);
 }
 
+/**
+ * Connect an unconnected knn graph (one in which mst returns an msf). The
+ * device buffers underlying the Graph_COO object are modified in-place.
+ * @tparam value_idx index type
+ * @tparam value_t floating-point value type
+ * @param[in] handle raft handle
+ * @param[in] X original dense data on host memory from which knn graph was constructed
+ * @param[inout] msf edge list containing the mst result
+ * @param[in] m number of rows in X
+ * @param[in] n number of columns in X
+ * @param[in] n_components number of components in color
+ * @param[inout] color the color labels array returned from the mst invocation
+ * @return updated MST edge list
+ */
+template <typename value_idx, typename value_t>
+void connect_knn_graph(
+  raft::resources const& handle,
+  const value_t* X,
+  raft::sparse::solver::Graph_COO<value_idx, value_idx, value_t>& msf,
+  size_t m,
+  size_t n,
+  int n_components,
+  value_idx* color,
+  cuvs::distance::DistanceType metric = cuvs::distance::DistanceType::L2SqrtExpanded)
+{
+  auto stream = raft::resource::get_cuda_stream(handle);
+
+  // Copy color array from device to host
+  std::vector<value_idx> h_color(m);
+  raft::copy(h_color.data(), color, m, stream);
+  raft::resource::sync_stream(handle, stream);
+
+  std::unordered_map<value_idx, value_idx> color_remap;
+  value_idx new_label = 0;
+
+  // Build remapping table so that colors are compact integers (i.e. consecutive colors)
+  for (size_t i = 0; i < m; ++i) {
+    if (color_remap.find(h_color[i]) == color_remap.end()) {
+      color_remap[h_color[i]] = new_label++;
+    }
+  }
+
+  for (size_t i = 0; i < m; ++i) {
+    h_color[i] = color_remap[h_color[i]];
+  }
```

Comment:
If my understanding of this operation is correct, could you use this primitive? https://github.com/rapidsai/raft/blob/9abfca9a3fc2fe7628775bc3c1b1b2cb247302c2/cpp/include/raft/label/classlabels.cuh#L113-L117

Reply:
Thanks! didn't know we had this